File: packages/ai_config/catalog/snapshot.py
import json
from typing import Dict


def build_catalog_snapshot(catalog: Dict) -> str:
    stable = {
        **catalog,
        "providers": sorted(catalog.get("providers", []), key=lambda p: p["id"]),
        "models": sorted(catalog.get("models", []), key=lambda m: (m.get("provider_id", ""), m["id"])),
    }
    return json.dumps(stable, indent=2, sort_keys=True)

File: packages/ai-config/src/catalog/snapshot.ts
import type { Catalog } from '../domain/catalog.js';

export function buildCatalogSnapshot(catalog: Catalog): string {
  const stable = {
    ...catalog,
    providers: [...catalog.providers].sort((a, b) => a.id.localeCompare(b.id)),
    models: [...catalog.models].sort((a, b) => {
      const byProvider = (a.providerId ?? '').localeCompare(b.providerId ?? '');
      return byProvider !== 0 ? byProvider : a.id.localeCompare(b.id);
    }),
  };

  return JSON.stringify(stable, null, 2);
}
